hbase-balancer/src/main/java/org/apache/hadoop/hbase/master/balancer/OverallSkewChecker.java

```diff
@@ -0,0 +1,109 @@
+ * groups or cluster, we hope the balancer can execute as soon as possible, but the
+ * StochasticLoadBalancer may need a lot of time to compute costs.
+ */
+@InterfaceAudience.Private
+class OverallSkewChecker implements StochasticLoadBalancer.ClusterStatusBalanceChecker {
+  private static final Logger LOG = LoggerFactory.getLogger(OverallSkewChecker.class);
+
+  private static final String BALANCER_OVERALL_SKEW_PERCENT_KEY =
+    "hbase.master.balancer.overall.skew.percentage";
+
+  private Configuration conf;
+
+  public OverallSkewChecker(Configuration conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public boolean
+    needsCoarseBalance(Map<TableName, Map<ServerName, List<RegionInfo>>> loadOfAllTable) {
+    if (this.conf == null) {
+      LOG.warn("Configuration should not be null");
+      return false;
+    }
+    if (isClusterSkew(loadOfAllTable)) {
+      LOG.info("Balancer checker found cluster is skew, will perform force balance, "
+        + "table count is {}", loadOfAllTable.size());
+      return true;
+    }
+    LOG.debug("Balancer checked cluster, it is not skew now, tables count is {}",
+      loadOfAllTable.size());
+    return false;
+  }
+
+  public boolean isClusterSkew(Map<TableName, Map<ServerName, List<RegionInfo>>> loadOfAllTable) {
+    if (loadOfAllTable == null || loadOfAllTable.isEmpty()) {
+      return false;
+    }
+    int skewCount = 0;
+    for (Map.Entry<TableName, Map<ServerName, List<RegionInfo>>> entry : loadOfAllTable
+      .entrySet()) {
+      if (isSkew(entry.getValue())) {
+        LOG.info("Table: " + entry.getKey().getNameAsString() + " regions count is skew");
+        if (LOG.isDebugEnabled()) {
+          StringBuilder sb = new StringBuilder(
+            "Table: " + entry.getKey().getNameAsString() + " regions distribution is: ");
+          for (Map.Entry<ServerName, List<RegionInfo>> regionDistribution : entry.getValue()
+            .entrySet()) {
+            sb.append(regionDistribution.getKey().getServerName()).append(" count=")
+              .append(regionDistribution.getValue().size()).append(":");
+            sb.append(regionDistribution.getValue());
+          }
+          LOG.debug(sb.toString());
+        }
+        skewCount++;
+      }
+    }
+    LOG.info("Cluster is skew, skew table count={}, load table count={}", skewCount,
+      loadOfAllTable.size());
+    return skewCount > 0 && (double) skewCount / loadOfAllTable.size()
+        > conf.getDouble(BALANCER_OVERALL_SKEW_PERCENT_KEY, 0.5);
+  }
+
+  private boolean isSkew(Map<ServerName, List<RegionInfo>> regions) {
+    if (regions == null || regions.isEmpty()) {
+      return false;
+    }
+    int max = 0;
+    int min = Integer.MAX_VALUE;
+    for (Map.Entry<ServerName, List<RegionInfo>> entry : regions.entrySet()) {
+      int count = entry.getValue() == null ? 0 : entry.getValue().size();
+      max = Math.max(max, count);
+      min = Math.min(min, count);
+    }
+    return max - min > 2;
```

Comment:
what is 2 here? this should be config driven as well? or based on some variance across servers?